### ISSUE TITLE
A question/suggestion on training using multiple tasks...

### ISSUE THREAD

**Issue by sjscotti** [facebookresearch/ParlAI#4336]
**Is your feature request related to a problem? Please describe.**

Hi
I have been using my own corpus to fine tune blenderbot 2 and I have tried doing it a couple of ways.

1.  just using my own corpus as the task and minimizing the loss
2. using my own corpus as a task, adding the Blended Skill Talk task, and minimizing the loss for these two tasks (I saw this as a suggestion in another issue to avoid overtraining)

For #1, I can get the training loss to almost zero, but the when doing an interactive chat using dialog that isn't similar to the training set I get terrible responses (often just repeating a training set response)

For #2, I get better results with interactive chat but it is still not as good as I had hoped for.  Looking at the validation during training I find that loss from just my corpus is primarily the one being minimized while the loss from Blended Skill Talk is increasing somewhat during the training (probably because of the `multitask-weights` I have selected). Looking at these individual task losses vs total loss, the total loss appears to be the average of the loss from the two tasks (I used `--multitask-weights 1, 1`, the documentation for this option says it is  `the probability of drawing the task in multitask case`, but from my experience that isn't what the parameter does since the ratio of the episodes isn't 1 to 1).  So it appears that the loss being minimized is just an average of the individual task losses - right?  

In searching  the web on multitask training, I found this interesting paper (https://paperswithcode.com/paper/multi-task-learning-as-multi-objective) that makes the point that just using a weighted average of different training tasks seldom gives a good trade off between the tasks, and makes the point what is really desired is a Pareto optimum of the losses of the individual tasks.

**Describe the solution you'd like**

The paper mentioned above provides an method to implement this solution that it claims is about the same computational cost as the weighted average approach.  It seems this would be a good method to implement in Parlai.  Could that be added to the list of future features? (However, see comments I later added below in **Additional context**)

**Describe alternatives you've considered**

**Additional context**
I noticed the paper mentioned says it deals with the case where the parameters being trained can be divided into shared and unshared groups.  This probably eliminates their method from use with training for blenderbot 2 models, but I think their point of needing a way of finding a better tradeoff (maybe a way to tune the `multitask-weights`?)  is still valid.

**Comment by mojtaba-komeili** [comment]
How big is your custom dataset and how diverse are the examples? Are you using any regularization (eg early stoppage)? It seems to me that you are overfitting on your custom dataset. If `multitask-weights` are 1 and your dataset is 1/100 of the size of BST it means that your examples are 100 time more likely to be shown tot he model than any example in BST.
As for your feature suggestion, if you think this feature provide benefits for all ParlAI users we welcome your contribution by forking our repository and submitting a PR.

**Comment by sjscotti** [comment]
@mojtaba-komeili 
My custom dataset is about 900 training exchanges while BST's `train.txt` is about 29000 exchanges, so yes, BST is much larger.  One point of confusion I have is that with `--multitask-weights 1, 1` , I still have several times more training exchanges coming from BST than from my custom dataset.  So it is not working like the documentation says since I would expect to see about the same number of exchanges coming from each set (as you suggested above).  So the actual application of multitask-weights does not appear to be `the probability of drawing the task in multitask case`, and it is also not how much each task's loss is weighted in computing the total loss, since total loss is just the arithmetic average of each task's loss (in my case of 2 task anyway) regardless of the ratio of `multitask-weights`.  So a good explanation of what `multitask-weights` is actually doing would be very helpful so it can be better used.

Regarding developing a parlai feature that does more than just average multiple tasks, but does a true multiobjective optimization... that is a development effort beyond my abilities.  It involves both developing a (or adapting someone else's) theoretical basis, and then developing a code implementation.  I am too much of a newbie in both of these areas to make such a contribution.

**Comment by stephenroller** [comment]
Your #1 probably amounts to simple overfitting.

I am supportive of a smarter strategy for optimization than task averaging. It's been a dream to implement one of these methods for over a year but it's always punted for something higher priority. Would you be interested in implementing it?

**Comment by stephenroller** [comment]
Regarding the ratio: MTW is calculated over episodes, not examples. So if your 900 utterances are all single turn, and BST is like 6-10 turns, I would expect BST data to show up 6-10x more often in practice.

**Comment by mojtaba-komeili** [comment]
> @mojtaba-komeili My custom dataset is about 900 training exchanges while BST's `train.txt` is about 29000 exchanges, so yes, BST is much larger. One point of confusion I have is that with `--multitask-weights 1, 1` , I still have several times more training exchanges coming from BST than from my custom dataset. So it is not working like the documentation says since I would expect to see about the same number of exchanges coming from each set (as you suggested above). So the actual application of multitask-weights does not appear to be `the probability of drawing the task in multitask case`, and it is also not how much each task's loss is weighted in computing the total loss, since total loss is just the arithmetic average of each task's loss (in my case of 2 tasks anyway) regardless of the ratio of `multitask-weights`. So a good explanation of what `multitask-weights` is actually doing would be very helpful so it can be better used.
> 
> Regarding early stopping to regularize the training. I agree that would be a good thing if I had a good criteria for when to stop. As I mentioned BST loss in actually increasing but the loss from my dataset is decreasing faster so the average of the two is decreasing. Is BST increasing a good thing? How should `multitask-weights` be used to affect this training process? And what do I use for a signal that I should stop the training?
> 
> Regarding developing a parlai feature that does more than just average the loss of multiple tasks, but does a true multiobjective optimization... that is a development effort beyond my abilities. It involves both developing a (or adapting someone else's) theoretical basis, and then developing a code implementation. I am too much of a newbie in both of these areas to make such a contribution.

I can think of two potential solutions that may help:
1. BlenderBot is already fin-tuned on BST and your dataset is already small. So, just fine-tune on your dataset.
2. Run the early stop validation only on your custom dataset. One was to do that is defining a custom metric in your task with a unique name (it is really easy to do in ParlAI), then use that as your `--validation-metric`.

**Comment by sjscotti** [comment]
> Regarding the ratio: MTW is calculated over episodes, not examples. So if your 900 utterances are all single turn, and BST is like 6-10 turns, I would expect BST data to show up 6-10x more often in practice.

Thanks @stephenroller 
That clears up my confusion.

**Comment by sjscotti** [comment]
> Your #1 probably amounts to simple overfitting.
> 
> I am supportive of a smarter strategy for optimization than task averaging. It's been a dream to implement one of these methods for over a year but it's always punted for something higher priority. Would you be interested in implementing it?

Hi @stephenroller 
I haven’t done an exhaustive search, but this repo and the papers it references seems like a possible approach to use to determine the best weighting of different task losses.

https://github.com/Mikoto10032/AutomaticWeightedLoss

But did you have a specific approach in mind?

**Comment by sjscotti** [comment]
> > @mojtaba-komeili My custom dataset is about 900 training exchanges while BST's `train.txt` is about 29000 exchanges, so yes, BST is much larger. One point of confusion I have is that with `--multitask-weights 1, 1` , I still have several times more training exchanges coming from BST than from my custom dataset. So it is not working like the documentation says since I would expect to see about the same number of exchanges coming from each set (as you suggested above). So the actual application of multitask-weights does not appear to be `the probability of drawing the task in multitask case`, and it is also not how much each task's loss is weighted in computing the total loss, since total loss is just the arithmetic average of each task's loss (in my case of 2 tasks anyway) regardless of the ratio of `multitask-weights`. So a good explanation of what `multitask-weights` is actually doing would be very helpful so it can be better used.
> > Regarding early stopping to regularize the training. I agree that would be a good thing if I had a good criteria for when to stop. As I mentioned BST loss in actually increasing but the loss from my dataset is decreasing faster so the average of the two is decreasing. Is BST increasing a good thing? How should `multitask-weights` be used to affect this training process? And what do I use for a signal that I should stop the training?
> > Regarding developing a parlai feature that does more than just average the loss of multiple tasks, but does a true multiobjective optimization... that is a development effort beyond my abilities. It involves both developing a (or adapting someone else's) theoretical basis, and then developing a code implementation. I am too much of a newbie in both of these areas to make such a contribution.
> 
> I can think of two potential solutions that may help:
> 
>     1. BlenderBot is already fin-tuned on BST and your dataset is already small. So, just fine-tune on your dataset.
> 
>     2. Run the early stop validation only on your custom dataset. One was to do that is defining a custom metric in your task with a unique name (it is really easy to do in ParlAI), then use that as your `--validation-metric`.

Hi @mojtaba-komeili 
Thanks for the suggestion.  I already have my task's dataset broken into a training dataset and validation dataset, and I am using the option `--evaltask` so that my validation dataset loss is computed for every step that does a validation computation.  (I am also using BST's validation dataset so I can see that its loss in increasing somewhat during training where BST is one of the training tasks, which has concerned me).   Four your solution, do  you have a criteria suggestion for selecting the early stopping point during the training?
Thanks!

**Comment by mojtaba-komeili** [comment]
Why don't you just keep your custom task in your `--evaltask`. Don't use BST if your final goal is using model on your own task. We know it is already fine-tuned on BST and your custom task is small. So, just fine-tune it slightly more on your own task. By overfitting on your own task you are deviating that from BST too.

**Comment by sjscotti** [comment]
Hi @mojtaba-komeili (and @stephenroller)

I already have almost a week of training time invested in training using both BST and my custom task, so I’d like to continue along that path.  From what I understand, including BST as an additional training task will reduce the tendency to overfit the model to my custom task.  I can do early stopping in either case, but I still need to determine when to stop.  Right now I am plotting the curves for total validation loss, BST validation loss, and custom task validation loss over the training with about 4 or 5 validation runs per epoch.  BST validation loss is slowly increasing, and my custom task loss and total loss are mostly decreasing.  When either of them show an increase, I cut the learning rate in half and continue (I’ve made some mods that allow me to do this with user input so I don’t have to restart the training).  Does this seem like a reasonable approach?

**Comment by sjscotti** [comment]
Hi @mojtaba-komeili and @stephenroller 
I was wondering if you could take a look at the validation loss history for my blenderbot2 400M fine tuning shown in the chart below, and give me your thoughts on which checkpoint (the symbols filled with red) would be the best trained checkpoint that avoids over training.  I am continuing the training with periodic reductions in learning rate, using the criteria I described above, in the hopes that the loss trends improve, but I don't think the trends will change.
Thanks!

![image](https://user-images.githubusercontent.com/44651388/152651103-53ca6211-8716-4d37-a1be-cebd12270d7d.png)

**Comment by stephenroller** [comment]
If BST loss is going up that much, you're probably only hurting the bot. You're definitely overfitting.

**Comment by sjscotti** [comment]
Just wanted to provide an update on what seemed to fix the problem I was having fine tuning BB2 400M with my own custom task.  I tried several things to improve the results for training using my custom task while also including the BST task.  Two things that helped a lot were: 1) using a lower maximum plateau learning rate (after a linear warmup) of 0.00005 using an Adam optimizer variant.  I was previously using a learning rate more than an order of magnitude higher.  Maybe that larger learning rate have been OK for initial training, but for fine tuning it just killed the bot as @stephenroller observed.  The learning rate was later reduced to help with convergence, but most of the convergence happened during the original plateau.  2) changing the `multitask-weights` so that, on the average, there were about the same number of executions (`exs`) in a step for each of the 2 training tasks.  Previously, for an `--update-freq `of 32 , only 2 or 3 `exs` were for my custom task and 29 or 30 were BST `exs`.    The BST loss stayed to within 1% of the initial value throughout the fine-tuning.  Fine tuning with these changes only took a few epochs and could be done in about a day on my desktop.
